Add Snowflake DQ Operators

denimalpaca opened the pull request:
Add three new Snowflake operators based on SQL Checks

The SnowflakeCheckOperator, SnowflakeValueCheckOperator, and
SnowflakeIntervalCheckOperators are added as subclasses of their respective
SQL Operators. These additions follow the conventions set in the BigQueryOperators
subclassing from the same SQL_CheckOperators.

closes: #17694

denimalpaca committed:
Add three new Snowflake operators based on SQL Checks

The SnowflakeCheckOperator, SnowflakeValueCheckOperator, and
SnowflakeIntervalCheckOperators are added as subclasses of their respective
SQL Operators. These additions follow the conventions set in the BigQueryOperators
subclassing from the same SQL_CheckOperators. This decision was made despite the
existing SnowflakeOperator, as subclassing both SnowflakeOperator and a SQL_CheckOperator
caused a couple issues:
  1. Both operators have an execute method already, so some precaution would be
	 necessary to choose the correct one.
  2. An issue with the `sql` parameter occured, where if it was given as a kwarg (sql=sql),
	 then the BaseOperator init() would through an exception for missing args, as if
	 it weren't being added to **kwargs correctly.

denimalpaca committed:
Create tests for new Snowflake Operators

Added three new tests, one for each new Snowflake Operator, in the same
style as the BigQuery check operators. These tests ensure the hook is properly
called. Not sure how helpful that really is, but it's coverage, and I have used
these new operators (well, two of them) in an actual DAG with a live connection,
and they're working.

The snowflake.py file was touched to fix the method name I changed when
trying other solutions, it was restored
to make the SnowflakeOperator pass its test again.

denimalpaca committed:
Merge branch 'apache:main' into snowflake_dq_operators

mik-laj commented on airflow/providers/snowflake/operators/snowflake.py:
```diff
@@ -125,3 +126,299 @@ def execute(self, context: Any) -> None:
+    :type sql: Can receive a str representing a sql statement,
+        a list of str (sql statements), or reference to a template file.
+        Template reference are recognized by str ending in '.sql'
+    :param use_legacy_sql: Whether to use legacy SQL (true)
```
Is it used?

mik-laj commented on airflow/providers/snowflake/operators/snowflake.py:
```diff
@@ -125,3 +126,299 @@ def execute(self, context: Any) -> None:
+
+    :param sql: the sql to be executed
+    :type sql: str
+    :param use_legacy_sql: Whether to use legacy SQL (true)
```
Snowflake have only one SQL variant. BigQuery support legacy SQL and ZetaSQL.

denimalpaca replied:
Ah, no. Mistakenly added as I used the BigQuery operator as a model.

denimalpaca committed:
Remove legacy SQL parameters (17741)

This parameter was mistakenly added.

mik-laj commented on airflow/providers/snowflake/operators/snowflake.py:
```diff
@@ -125,3 +126,284 @@ def execute(self, context: Any) -> None:
+    def get_db_hook(self) -> SnowflakeHook:
+        """
+        Create and return SnowflakeHook.
+        :return: a SnowflakeHook instance.
```
```suggestion

        :return: a SnowflakeHook instance.
```
this is needed to render the documentation properly.

mik-laj committed:
Update airflow/providers/snowflake/operators/snowflake.py

JavierLopezT commented on airflow/providers/snowflake/operators/snowflake.py:
```diff
@@ -125,3 +126,285 @@ def execute(self, context: Any) -> None:
             return execution_info
+
+
+class _SnowflakeDbHookMixin:
```
Why a class and not just a method inside the SnowflakeCheckOperator?

potiuk replied:
Good point

denimalpaca replied:
Trying to be more DRY. Originally I tried making the `Check` classes inherit from just the `SnowflakeOperator`, but it resulted in an error in the constructors.

JavierLopezT replied:
DRY philosophy is great. However, my point was that you don't need a class at all and you can get rid of the class extra layer of abstraction. My comment is inspired by https://www.youtube.com/watch?v=o9pEzgHorH0

Using the same class too in the SnowflakeOperator has been quite an improvement, but I think you can still have just a function in the file, and you can just call it from the operators. Anyway, it's not a big deal and the PR is already approved, so up to u.

potiuk replied:
I wached the presentation :). Entertaining and I mostly agree with it  (not everything :) 

But yeah - in  this case using mixin is definitely over-the-top.

potiuk replied:
Would you mind changing that @denimalpaca ?

denimalpaca replied:
Will have this fixed Tuesday, looks like with the way BaseSQLOperator works, even the function without a class might not be necessary.

JavierLopezT commented on airflow/providers/snowflake/operators/snowflake.py:
```diff
@@ -125,3 +126,285 @@ def execute(self, context: Any) -> None:
+    :param snowflake_conn_id: Reference to
+        :ref:`Snowflake connection id<howto/connection:snowflake>`
+    :type snowflake_conn_id: str
+    :param autocommit: if True, each command is automatically committed.
```
Given that the the sqls are meant to be SELECT statements, does it make sense to have this param?

potiuk replied:
The can be also UPDATE statements

JavierLopezT commented on airflow/providers/snowflake/operators/snowflake.py:
```diff
@@ -125,3 +126,285 @@ def execute(self, context: Any) -> None:
+
+class SnowflakeValueCheckOperator(_SnowflakeDbHookMixin, SQLValueCheckOperator):
+    """
+    Performs a simple value check using sql code.
```
I know that this is copied from https://github.com/apache/airflow/blob/main/airflow/operators/sql.py#L149, but I think it's unclear what the operator really is. I think adding something like 'it checks against a value that you define' would be really useful

JavierLopezT commented on airflow/providers/snowflake/operators/snowflake.py:
```diff
@@ -125,3 +126,285 @@ def execute(self, context: Any) -> None:
+    def __init__(
+        self,
+        *,
+        sql: Any,
```
Is it really Any? Or union str and list?

potiuk replied:
Good point too.

denimalpaca replied:
Actually just `str` according to the [SQL Operator](https://airflow.apache.org/docs/apache-airflow/stable/_modules/airflow/operators/sql.html#SQLValueCheckOperator) it inherits from.

JavierLopezT commented on airflow/providers/snowflake/operators/snowflake.py:
```diff
@@ -125,3 +126,285 @@ def execute(self, context: Any) -> None:
+
+    :param table: the table name
+    :type table: str
+    :param days_back: number of days between ds and the ds we want to check
```
If this is here, I guess `tolerance` and `pass_value` should be in the docstring of `SnowflakeValueCheckOperator`. It seems that for Sphinx is not necessary https://stackoverflow.com/a/2025599/9621172 but certainly it really helps to have them here if you are just reading the code

denimalpaca replied:
Agreed. Those docstrings aren't even in the SQL Operators (and should be).

denimalpaca committed:
Merge branch 'apache:main' into snowflake_dq_operators

denimalpaca committed:
Make code more DRY by having all classes use hook mixin class

denimalpaca committed:
Update SnowflakeValueCheckOperator docstring

denimalpaca committed:
Merge branch 'apache:main' into snowflake_dq_operators

denimalpaca committed:
Modify SnowflakeOperator to use get_db_hook and update test

In a previous commit, the get_hook() method was removed from
SnowflakeOperator, and the mixin class containing this functionality
was used instead. This commit finishes updating the necessary name changes
in the original class and test.

denimalpaca committed:
Merge branch 'apache:main' into snowflake_dq_operators

denimalpaca committed:
Removes Snowflake Mixin Class

The SnowflakeOperatorMixin class was deemed an unncessary abstraction,
so the get_db_hook() method is used as a standalone function called by
the Operators directly. Tests are updated to reflect this change.
